### PR TITLE
Draw duplicate pips of a separate suits differently

### LINKED ISSUE #2251: Modify the second suit pip of double almost same suit variants
E.g. "Light Pink and Pink", or "Muddy Rainbow and Cocoa Rainbow" have this problem: the cards look very similar, and can be confused. One workaround is colorblind mode, because it always shows the letter in corner.

My proposal which works for non-colorblind mode too is to change the pip in some way to make it distinguishable form the base pip. Maybe add some pattern around the pip.

### PR DESCRIPTION
**Pull request by DarthGandalf**
Fix #2251

### CODE REVIEW

File: .cspell.json [hunk 1]
```diff
@@ -141,6 +141,7 @@
         "kadda",
         "keldon",
         "keldon's",
+	"keyable",
```
Comment (Zamiell) [inline .cspell.json:144]: looks like you manually inserted this instead of using the cspell took, because the whitespace is wrong

File: client/src/game/ui/drawPip.ts [hunk 2]
```diff
@@ -3373,6 +3375,47 @@ export default function drawPip(
...
+      for (let i = 0; i < colors.length; ++i) {
+        grad.addColorStop(i / (colors.length - 1), colors[i]);
+      }
+      ctx.fillStyle = grad;
+    } else {
+      ctx.fillStyle = suit.fill;
+    }
+    // Workaround for Konva.Shape's Context not supporting evenodd
+    // This is used in e.g. Empathy view
+    interface Keyable {
```
Comment (Zamiell) [inline client/src/game/ui/drawPip.ts:3401]: i think this is better written as
```ts
type Keyable = Record<string, any>
```
correct me if i am wrong

File: .cspell.json (hunk 1 above)
Comment (DarthGandalf) [reply .cspell.json]: I didn't know cspel can edit this file. Fixed.

File: client/src/game/ui/drawPip.ts (hunk 2 above)
Comment (DarthGandalf) [reply client/src/game/ui/drawPip.ts]: You're right. I just don't know any typescript

File: client/src/game/ui/drawPip.ts [hunk 3]
```diff
@@ -3373,6 +3375,45 @@ export default function drawPip(
...
+      const colors = suit.fillColors;
+      const grad = ctx.createLinearGradient(0, -100, 30, 100);
+      for (let i = 0; i < colors.length; ++i) {
+        grad.addColorStop(i / (colors.length - 1), colors[i]);
+      }
+      ctx.fillStyle = grad;
+    } else {
+      ctx.fillStyle = suit.fill;
+    }
+    // Workaround for Konva.Shape's Context not supporting evenodd
```
Comment (DarthGandalf) [inline client/src/game/ui/drawPip.ts:3399]: Konva fixed this in version 8, but this project still uses version 3.4.1

**Comment by Zamiell** [comment]
cspell is still bugged, you added a newline at the end

**Comment by DarthGandalf** [comment]
Fixed everything except for "unknown", it needs "any".

      TS2352: Conversion of type 'CanvasRenderingContext2D' to type 'Record<string, unknown>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Index signature is missing in type 'CanvasRenderingContext2D'.

**Comment by Zamiell** [comment]
yeah, so cast it as unknown first exactly like the error message describes

**Comment by DarthGandalf** [comment]
Right. I was trying to cast the wrong variable, so it didn't work

File: client/src/game/ui/drawPip.ts [hunk 4]
```diff
@@ -3373,6 +3375,44 @@ export default function drawPip(
...
+    // This is used in e.g. Empathy view
+    type Keyable = Record<string, unknown>;
+    // eslint-disable-next-line
```
Comment (Zamiell) [inline client/src/game/ui/drawPip.ts:3402]: why is this here?

Comment (DarthGandalf) [reply client/src/game/ui/drawPip.ts:3402]: 3402:50  error  ["_context"] is better written in dot notation                                                       @typescript-eslint/dot-notation

Comment (DarthGandalf) [reply client/src/game/ui/drawPip.ts:3402]: But the dot notation also shows error, because that type isn't supposed to have such field

**Comment by DarthGandalf** [comment]
The whole Keyable is there only to workaround this issue with konva<8.0.0. Do you see some solution for this which doesn't involve disabling eslint?

Ideally, probably, update konva to >=8.0.0

**Comment by Zamiell** [comment]
upgrading konva is complicated, when we tried last time there were some bugs, but if you take the commit where we upgraded and then fix the bugs from there, i would accept a PR for that

**Comment by DarthGandalf** [comment]
Yeah, I'm not up to that task yet.

**Comment by DarthGandalf** [comment]
Lint error is better than the compilation error, at least

**Comment by Zamiell** [comment]
see my commit on how to model the data properly. its kind of a hack because the real type of the argument is the KonvaCanvasType thing but I don't know what it is

**Comment by Zamiell** [comment]
thanks and nice work

**Comment by DarthGandalf** [comment]
It looks like some internal Konva's type

**Comment by Zamiell** [comment]
its some Konva type but presumably it is exposed for use in the typescript definitions
